### PR TITLE
Added NotificationWillBeSent plugin's hook

### PR DESCRIPTION
**Pull request by rfoyard**
#### Summary
Added a new plugin's hook `NotificationWillBeSent`.  
This hook's arguments are the post triggering the notification and the list of users to be notified.  
The return value is the modified list of users to be notified.

#### Ticket Link
JIRA: [MM-19202](https://mattermost.atlassian.net/browse/MM-19202)
Issue: #12624

#### Release Note
```release-note
Added NotificationWillBeSent plugin hook 
```

### CODE REVIEW

**Comment by cpoile** [comment]
@rfoyard Wow, thanks, this looks great! 

To help the review process, can you point out which functions are brand new (ie, you are adding for this PR) and ones that might be refactored from other places in the code (ie, not new code). It's tough to tell in this PR, and the reviewers might not be familiar with this part of the codebase.

That would help us a lot!

**Comment by rfoyard** [comment]
@cpoile Sure, the refactored code has been moved from `app/notification.go` to `model/notification.go` (and the according `_test.go` file). All of this code is for the `ExplicitMentions` structure, methods and helper functions. This was necessary because the plugin needed access to these and it incurred a cycling dependency between `app/` and `plugin/`.

The new code has been added to:
 - [app/notification.go](https://github.com/mattermost/mattermost-server/pull/13775/files#diff-88770258b63cdda80e0dfef721631f8dR126)
 - [app/plugin_hooks_test.go](https://github.com/mattermost/mattermost-server/pull/13775/files#diff-d74c326c126d2e2d5507081477ab8a5dR665)
 - [plugin/hooks.go](https://github.com/mattermost/mattermost-server/pull/13775/files#diff-81d1bc59cb074b28cbda79d37dca9f56)
 - [plugin/plugintest/hooks.go](https://github.com/mattermost/mattermost-server/pull/13775/files#diff-eadcbf8ece04711fa69be6ded0a87732R155)

The file [plugin/client_rpc_generated.go](https://github.com/mattermost/mattermost-server/pull/13775/files#diff-d74c326c126d2e2d5507081477ab8a5dR665) is automatically generated using `make pluginapi`

File: plugin/hooks.go
Comment (hanzei) [inline plugin/hooks.go]: What happens if a pluin returns `nil`?

Comment (hanzei) [inline plugin/hooks.go]: I'm wondering if we also should add a `NotificationHasBeSent` while touching this code piece. I should not be a lot of effort. @rfoyard Thoughts?

File: model/notification.go
Comment (hanzei) [inline model/notification.go]: Is there any use case for overwriting this property?

File: model/notification.go
Comment (hanzei) [inline model/notification.go]: Same question here.

File: app/plugin_hooks_test.go
Comment (hanzei) [inline app/plugin_hooks_test.go]: Nit:
```suggestion
	assert.Len(t, newMentions.Mentions, 2)
```

File: plugin/hooks.go
Comment (rfoyard) [reply plugin/hooks.go]: This is handled in the `plugin/plugintest/hooks.go` file, [here](https://github.com/mattermost/mattermost-server/pull/13775/files#diff-eadcbf8ece04711fa69be6ded0a87732R163)  
If the call returns `nil`, the original `mentions` arguments is returned back.

File: model/notification.go
Comment (rfoyard) [reply model/notification.go]: This is a good question but I'm not the author of this code, I'll try to look at the source PR for this change.

File: model/notification.go
Comment (rfoyard) [reply model/notification.go]: The same answer here ;)

File: plugin/hooks.go
Comment (rfoyard) [reply plugin/hooks.go]: I'm not sure I can find a use case for this, but it should be possible.  
The thing is there are different type of notification that can be sent, a user can receive an email and a push notification and an ephemeral post (at the same time !).  
They all are using a different code path and I'm not sure how to deduplicate them.  
Do you see any use case for such a callback ?

Comment (hanzei) [reply plugin/hooks.go]: Are you sure? As far as I understand the code, this layer just make sure the results from the plugin as propagated to the `app` layer. There are other hooks that allow `nil` as a return value for there hooks.

Comment (hanzei) [reply plugin/hooks.go]: There might be the use case to mirror notification to an external application .

But this is not a blocking request. If it's hard to implement, just leave it out.

File: model/notification.go
Comment (hanzei) [reply model/notification.go]: Just from looking at the code I can see that only `model` is using this property. Maybe a getter is enough for plugins? Then it doesn't have to be exported.

File: plugin/hooks.go
Comment (rfoyard) [reply plugin/hooks.go]: Well, this function `NotificationWillBeSent(c *Context, post *model.Post, mentions *model.ExplicitMentions) *model.ExplicitMentions` is to be implemented by the plugin.  
The implementation in `plugin/plugintest/hooks.go` is where it's actually called from the server and it does handle the case of a nil value iirc.

File: model/notification.go
Comment (rfoyard) [reply model/notification.go]: `HereMentioned`is used in [app/notification.go](https://github.com/mattermost/mattermost-server/blob/3e3fcdf84502ac0a4652e1d6e0459e67ed5ac188/app/notification.go#L177) as well

File: model/notification.go
Comment (rfoyard) [reply model/notification.go]: `OtherPotentialMentions`is used in [app/notification.go](https://github.com/mattermost/mattermost-server/blob/3e3fcdf84502ac0a4652e1d6e0459e67ed5ac188/app/notification.go#L109) too

File: model/notification.go
Comment (hanzei) [reply model/notification.go]: But only to read the variable. A getter would be enough for `app`.

File: plugin/hooks.go
Comment (hanzei) [reply plugin/hooks.go]: I'm mostly concerned that this is an edge case that will cause problems with the code. Would you mind writing an integration test in https://github.com/mattermost/mattermost-server/tree/3e3fcdf84502ac0a4652e1d6e0459e67ed5ac188/app/plugin_api_tests to confirm the app runs fine if `nil` is returned? I suppose it has to be an `manual` one. See https://github.com/mattermost/mattermost-server/blob/3e3fcdf84502ac0a4652e1d6e0459e67ed5ac188/app/plugin_api_test.go#L827-L832 for more details.

Comment (cpoile) [reply plugin/hooks.go]: IMO, 4/5 make a separate ticket for that, since it would probably involve asking "who needs that hook, and why" and "would that hook run on every type of notification, or just x?" etc.

File: model/notification.go
Comment (rfoyard) [reply model/notification.go]: Same as before, It's not a big change actually, I'll fix it in the next commit.

File: model/notification.go
Comment (rfoyard) [reply model/notification.go]: It's not a big change actually, I'll fix it in the next commit.

File: plugin/hooks.go
Comment (rfoyard) [reply plugin/hooks.go]: I'll first fix the `app/notification.go` issue and will look into this for another commit.

File: app/notification.go
Comment (rfoyard) [reply app/notification.go]: I'll fix this in the next commit !

**Comment by jasonblais** [comment]
cc @hanzei @cpoile Wasn't sure on the status, looks like some changes have been made by @rfoyard since the last review

**Comment by hanzei** [comment]
@jasonblais This PR is on my review list. I will take a look this week.

File: plugin/hooks.go
Comment (hanzei) [reply plugin/hooks.go:164]: Could you, just to double chat we catch this edge case, add a test to `app/plugin_hooks_test.go` where `NotificationWillBeSent` returns `nil`? Thanks

**Comment by hanzei** [comment]
@rfoyard Heads up that there is a conflict to resolve. Let me know, if you have any questions.

**Comment by teresa-novoa** [comment]
@rfoyard friendly reminder about @hanzei comment on April 4: _Heads up that there is a conflict to resolve. Let me know, if you have any questions._ Thanks!

**Comment by teresa-novoa** [comment]
@hanzei @jasonblais where are we with this PR? Thanks!

**Comment by jasonblais** [comment]
Hi @rfoyard, apologies for the long review time for this PR. I'm wondering if you are still interested and open to helping with it - looks like it's very close to being completed!

We'd be happy to help with any questions :)

**Comment by rfoyard** [comment]
Sorry for the delay, I've been quite busy but I'll try to fix the conflicts next week.  
Let me know if it's ok for you :)

**Comment by teresa-novoa** [comment]
Thanks @rfoyard!

**Comment by jasonblais** [comment]
Hi @rfoyard, touching base to see if you'll have time to revisit this PR? If not yet, totally understand, just mostly reaching out to hear if you have any questions :)

**Comment by rfoyard** [comment]
Hi @jasonblais, the holidays are coming 😄 . I'll be able to find some time this month to look at it so you can keep this issue opened. Thanks !

**Comment by cpoile** [comment]
@rfoyard @jasonblais I think we will have to pause on this, or cancel it altogether (sorry!). I noticed in the R&D meeting that the UX team has a similar feature designed out, which would replace this one.

**Comment by jasonblais** [comment]
@cpoile Does the UX design impact the server plugin hook itself?

**Comment by cpoile** [comment]
@jasonblais I'm not 100% sure of course, but it looked like they want the feature to be a lot more comprehensive--e.g., a watched threads view, which means there will be a list of watched threads in the db, and that's what will drive the notifications. So it's more than what a plugin will be able to handle. But regardless, that's up to the apps team to decide how they want to architect it--it'll be out of Toolkit team's hands.

**Comment by jasonblais** [comment]
@cpoile I see.

I wonder if there are other use cases for this plugin hook beyond the 'watch/unwatch thread' feature, which other plugins could leverage. cc @aaronrothschild

**Comment by hanzei** [comment]
@jasonblais There are indeed other use cases like intercepting notifications and use a custom build notification system.

**Comment by jasonblais** [comment]
@hanzei Okay, I've assigned to Aaron for PM ownership

**Comment by jasonblais** [comment]
@aaronrothschild any next steps we can share with the contributor?

**Comment by aaronrothschild** [comment]
Did a quick read through here, and it seems like this code could be useful as @hanzei mentioned for implementing a custom notification system that fires of to another system for example.  

I'm not too sure about the Core teams plans for handling threading etc. but will defer to Hanzei and others on if merging this will overlap with that functionality.  

What is left to do on this feature?

**Comment by rfoyard** [comment]
@aaronrothschild @hanzei  Hey there ! Sorry I didn't answer sooner.
I'm still available to finish this PR but I'd like to know what's missing from it (Are the current features enough ? Do you need something more ? Some thoughts about tests ? )
Please let me know and I'll try to finish this asap !

**Comment by hanzei** [comment]
@rfoyard My only remaining request is in https://github.com/mattermost/mattermost-server/pull/13775#pullrequestreview-387687686. Otherwise the PR looks good.

And could you please resolve the conflicts?

**Review by hanzei — APPROVED** [review]
LGTM. Thanks for coming back to this PR!

File: plugin/hooks.go
Comment (hanzei) [inline plugin/hooks.go:217]: That need to be 5.31 now
```suggestion
	// Minimum server version: 5.31
```

**Comment by rfoyard** [comment]
> LGTM. Thanks for coming back to this PR!

Thanks, I'm glad we could make it work !

File: app/plugin_api_test.go
Comment (cpoile) [inline app/plugin_api_test.go:1836]: @hanzei , do you think we should have some other integration tests here? Perhaps testing a few of the notification types (at-user, at-channel, etc)?  
TBH, I'm not sure of the best way to do that from this test. Maybe you have some insight on how to do it?

Comment (hanzei) [reply app/plugin_api_test.go:1836]: Does the test in `app/plugin_hooks_test.go` cover your request? (OT: Maybe we should fall back on only using ` app/plugin_api_tests`)

Comment (cpoile) [reply app/plugin_api_test.go:1836]: @hanzei yes, that's good, thanks!

Up to you about using `app/plugin_api_tests`, I have no strong opinion

**Review by cpoile — APPROVED** [review]
@rfoyard Apologies about the delay and thrashing on this, but it looks great, thank you!

The PR needs to sync and merge some conflicts, and then it's done, good job!

**Comment by rfoyard** [comment]
> @rfoyard Apologies about the delay and thrashing on this, but it looks great, thank you!
> The PR needs to sync and merge some conflicts, and then it's done, good job!

No problem ! Thank you for the review !

**Comment by hanzei** [comment]
@DHaussermann This PR is best tested by implementing the hook into the demo plugin.

@rfoyard As a final request: Would you be interested in implementing this hook into https://github.com/mattermost/mattermost-plugin-demo/ and therefore enabling testing of it? It's totally fine if you don't want to jump on that. In that case a core developer will take care of that.

**Comment by emilyacook** [comment]
> @DHaussermann This PR is best tested by implementing the hook into the demo plugin.
> 
> @rfoyard As a final request: Would you be interested in implementing this hook into https://github.com/mattermost/mattermost-plugin-demo/ and therefore enabling testing of it? It's totally fine if you don't want to jump on that. In that case a core developer will take care of that.

@rfoyard gentle ping ^^

**Comment by isacikgoz** [comment]
Thanks for all the effort @rfoyard 🙏 I am going to close this PR as you [mentioned](https://github.com/mattermost/mattermost-server/issues/12624#issuecomment-801091666) that you are not going to continue working on this. I am sorry to hear that experience didn't go well for you. We can re-open this if someone wants to continue working on the PR.

**Comment by mickmister** [comment]
@hanzei Thoughts on queuing this for Demansol?

**Comment by hanzei** [comment]
@mickmister Sounds good to me :+1:

**Pull request closed without merging**
